### ISSUE TITLE
Location and Exception information

### ISSUE THREAD

**Issue by lp55** [boost-experimental/ut#236]
Hi,

   I'm working on the [ApprovalTests.cpp](https://github.com/approvals/ApprovalTests.cpp/) integration and I've ran into a couple issues:

- I need the filename from where the test is implemented (caller function). assertion_pass and assertion_fail have that information but not test_begin. As this information needs to be set before the test is actually run, test_begin needs to have that information. I've been looking at the UT code, but haven't quite figure out the best way to do it.

- In ApprovalTests.cpp, when a test fails it throws an exception which import information inside the exception (e.what()). So when a test throws an exception I would like to check it's contents (and possibility it's type), not simply see if the test threw an exception or not, as it's implemented now. What do you think it would be the best approach here. Simply to add a std::string to the exception struct and store the exception data there? Again not sure the best approach to change the code here.

    Any help would be appreciated. If you want to take a look at the integration between ApprovalTests.cpp and UT take a look at this [pull request](https://github.com/approvals/ApprovalTests.cpp/pull/58) at the UTApprovals.h file.

   Thanks!

**Comment by krzysztof-jusiak** [comment]
Hi @lp55,

* Adding source_location is a bit tricky to ""_test (it's not to test/should/given/when/then) because the UDL has a strict requirement of parameters which can be passed into it and makes it harder to pass source_location with a default value, anyway, let me experiment with it...

* Not sure whether I fully follow but your approach seems reasonable, or at least it's a good start :+1: It's KISS.

Let me dig in into the above and will check out your PR with any updates afterwards.

**Comment by lp55** [comment]
Hi @krzysztof-jusiak 

Yeah the source_location part is a bit tricker. Let me know how your experiments go.
For the exception one, I thought about it a bit more and I came out with something simple, but I think is ok as it uses std::exception as a base, which I think it's a fair presumption. I made this [pull request](https://github.com/boost-experimental/ut/pull/238). Take a look and see what you think.
